[정연주] - 폴더 구조 정리, 옵저버 리팩토링, 메인 영역 API 연결, 메인 영역 필터링 기능 추가, 캘린더 추가

yyeonzu opened the pull request:
## 완료 작업 목록
1) 전반적인 폴더 구조를 다음과 같이 정리했습니다.
  ```css
  src/
  ├── apis/
  ├── assets/
  ├── components/
  │   ├── header/
  │   ├── transaction-form/
  │   └── transaction-list/
  ├── constants/
  ├── observers/
  │   ├── headerObserver.js
  │   ├── transactionFormObservers.js
  │   └── TransactionListObserver.js
  ├── pages/
  ├── store/
  ├── styles/
  └── utils/
  ```
2) store, 옵저버와 관련한 부분을 리팩토링했습니다.
3) 메인 영역의 데이터에 대해 서버 API를 연결했습니다.(추가, 삭제, 수정) (json-server 패키지 활용)
4) 메인 영역의 필터링 기능을 구현했습니다.
- 추가, 수정, 삭제 연결
      <img src="https://github.com/user-attachments/assets/771fa5b9-b06b-4fb3-837c-9a8d463fb732" width="420"/>
- 필터링
      <img src="https://github.com/user-attachments/assets/681234f3-d377-4018-a5aa-a26f91d5a746" width="420"/>

5) 캘린더 기능을 구현했습니다.
    <img width="420" alt="image" src="https://github.com/user-attachments/assets/0ed6c029-0fc6-4978-a67f-3253e2e00464" />



## 주요 고민과 해결 과정

**1) 폴더 구조 정리** 
폴더 구조가 복잡해져 가독성이 떨어진다고 느껴 정리를 했습니다. components 내부에 세부 폴더를 만들어 각 컴포넌트를 관리하도록 설계했습니다. 하나의 컴포넌트 내부에서 controller, view, events를 파일로 분리하여 작성했습니다. 완전히 MVC 패턴을 구현했다고 생각하진 않지만.. 최대한 MVC 패턴과 유사하게 만들려고 노력했습니다.
<br/>

**2) 옵저버가 아닌 곳에서도 view를 불러오는 문제 해결**
기존에는 옵저버 패턴임에도 controller와 view 간의 의존성이 있었는데, 이를 최소화하기 위해 노력했습니다. "controller에서는 state만 변경"에 집중하여 controller와 view의 관계를 느슨하게 변경했습니다. 
최초의 view를 그리는 함수도 observer에서 처리하고자 했는데, 하다보니 코드의 복잡성이 높아지고 가독성은 떨어지는 것 같아.. 최초 1회의 렌더링만 controller에서 담당하도록 (스스로) 합의했습니다.
완성하고보니 observer에서 부분렌더링을 수행할 때 기존의 view 컴포넌트를 활용하지 못하는 부분이 많다는 것이 좀 아쉽습니다. ㅠㅠ
<br/>

**3) API 연결 과정에서 불필요하게 API를 한 번 더 요청하는 과정 삭제**
처음에는 중첩된 then을 사용해 서버에 post한 데이터를 다시 get해서 화면에 렌더링하는 과정을 거쳤는데, store를 활용해서 서버에 post를 성공할 경우 다음과 같이 store의 상태를 업데이트해 화면을 리렌더링하여 불필요한 API 호출을 줄였습니다. (종성님 코드를 참고했습니다!)

기존 코드
```javascript
function submitPost(nowItem) {
  addItem(nowItem)
    .then(() => {
      return getItems();
    })
    .then((fetchedItems) => {
      ItemsState.setItems(fetchedItems);
      FormState.resetFormState();
    })
    .catch((error) => {
      throw error;
    });
}
```
수정한 코드
```javascript
function submitPost(nowItem) {
  addItem(nowItem)
    .then((newItem) => {
      const prevItems = ItemsState.getItems();
      const updatedItems = [...prevItems, newItem];
      ItemsState.setItems(updatedItems);
      FormState.resetFormState();
    })
    .catch((error) => {
      throw error;
    });
}
```
<br/>

**4) 이벤트 위임 시 이벤트 실행 순서 때문에 발생하는 문제 해결**
하나의 컴포넌트에서 하나의 이벤트 리스너를 통해 이벤트를 관리하도록 설계했는데, 입출금 내역의 "click" 이벤트에는 삭제와 수정이 모두 있어서, 만약 수정 이벤트 처리를 먼저 할 경우 삭제 버튼을 클릭해도 수정으로 적용된다는 문제가 있었습니다. 이 때문에 더 깊은 요소의 이벤트 처리를 먼저 한 뒤, 넓은 요소의 이벤트를 처리했습니다.

그리고 이벤트 처리 함수를 분리하는 과정에서 다음과 같이 코드를 작성했는데,
```javascript
export function addTransactionListEvents() {
  addEvent({
    id: "transaction-list",
    event: "click",
    onEvent: (e) => {
      handleFilterList(e);
      handleDeleteItem(e);
      handleUpdateItem(e);
    },
  });
}
```

handleDeleteItem으로 삭제를 진행한 뒤에도 handleUpdateItem이 실행되는 버그가 있어 함수의 return을 boolean 값으로 받고 다음과 같이 수정했습니다. 너무 임시방편으로 해결한 것이 아닌가 싶어서.. 내일 조금 더 고민해보겠습니다.
```javascript
export function addTransactionListEvents() {
  addEvent({
    id: "transaction-list",
    event: "click",
    onEvent: (e) => {
      handleFilterList(e);
      if (handleDeleteItem(e)) return;
      handleUpdateItem(e);
    },
  });
}
```

**5) 옵저버가 중복해서 등록되는 문제 해결**

페이지 라우팅이 추가되면서 main -> calendar 으로 이동 후  상태를 변경했을 때, 옵저버가 존재하지 않는 main의 요소를 변경하려고 해서 에러가 발생하여 문제를 알게 되었습니다.

처음에는 옵저버를 구독해제 하지 않고, 렌더링 함수에서  해당 요소가 존재하는지 확인하는 코드를 추가했습니다. 
```javascript
export function renderComponent({ id, className, innerHTML }) {
  const content = document.getElementById(id);
  if (!content) return;
  content.className = className;
  content.innerHTML = innerHTML;
}
```
하지만....

<img width="1302" height="312" alt="image" src="https://github.com/user-attachments/assets/38f84bfc-f11b-4df3-93fe-39137cbe29f6" />

페이지(레이아웃)를 이동하고 다시 돌아올 때마다 옵저버를 등록하기 때문에 옵저버가 중복되어 등록된다는 것을 발견했습니다.
_(처음에는 5번 클릭했을 때 update가 5번 실행됐는데, 페이지를 n번 이동하면 *n 만큼 update 함수가 실행됨)_


페이지 변경에 앞서, 존재하는 모든 옵저버를 제거하고 페이지를 init하도록 코드를 수정했습니다.
```javascript
class LayoutObserver {
  update() {
    // 모든 옵저버 해제
    removeTransactionListObservers();
    removeTransactionFormObservers();
    removeCalendarObserver();

    const state = NavBarState.getNavBarState();
    switch (state) {
      case "메인":
        createMainLayout();
        break;
      case "캘린더":
        createCalendarLayout();
        break;
      default:
        createMainLayout();
    }
  }
}
```

결과
<img width="1318" height="264" alt="image" src="https://github.com/user-attachments/assets/6b8ee0ee-2dae-4374-8ca0-101161b5d03c" />

페이지 이동 후에도 update 함수가 한번씩만 실행되는 것을 확인할 수 있습니다. (상태 변경 5번 수행함)


**6) FormState의 날짜와 DateStae의 날짜 데이터 흐름을 한방향으로 지정**

기존에는 입력 폼의 date input에서 FormState와 DateState를 동시에 구독하여 결과를 반영했는데, 그렇게 하더라도 DateState의 변화가 반영되지는 않는다는 단점이 있었습니다.

이를 해결하기 위해 FormState만 구독하고, 대신 DateState가 변경되면 FormState 역시 변경되도록 데이터의 흐름을 단방향으로 지정했습니다.

기존 코드
```javasciprt
class DateInputObserver {
  constructor() {
    FormState.subscribe(this);
    DateState.subscribe(this);
  }

  update() {
    const input = document.getElementById("date-input");
    if (input) {
      input.value = FormState.getFormState().date;
    }
  }
}
```

개선한 코드
```javascript
class DateState extends Observable {
  constructor() {
    super();
    this.curDate = getDateYMD(new Date());
  }
  setDate(newDate) {
    this.curDate = newDate;
    FormState.setDate(newDate);
    this.notify();
  }
  getDate() {
    return this.curDate;
  }
}
```
```javascript
class DateInputObserver {
  constructor() {
    FormState.subscribe(this);
  }

  update() {
    const input = document.getElementById("date-input");
    if (input) {
      input.value = FormState.getFormState().date;
    }
  }
}
```

Commit:
refactor: 입출금내역 view 분리, 옵저버 등록

Commit:
fix: 클래스 이름 오타 수정

Commit:
style: 삭제 버튼 pointer 추가

Commit:
feat: 입출금내역 수정 기능 추가

Commit:
chore: gitignore 파일 추가

Commit:
chore: json-server 패키지 설치

Commit:
chore: 서버 더미 데이터 추가

Commit:
chore: api 주소 상수화

Commit:
feat: 서버 데이터 fetch 함수 추가

Commit:
chore: 기존의 store 삭제

Commit:
chore: store의 state에 init 추가
init 메서드를 통해 최초 화면 렌더링을 observer를 통해 할 수 있도록 설정

Commit:
chore: 폴더 구조 변경
components/header 내부에서 controller, view, event 함수를 따로 관리하도록 수정

Commit:
refactor: header 렌더링 구조 변경
- observer에서 처리하지 않고, init 함수에서 최초 렌더링 수행
- renderHeader 내부의 파라미터(관심사)는 renderheader 내부에서 선언

Commit:
refactor: state의 init 메서드 삭제

Commit:
refactor: transaction form 관련 폴더 구조 변경 반영

Commit:
Merge branch 'chore/cleaning-folder' into dev

Commit:
refactor: transaction list 관련 함수 폴더 구조 변경

Commit:
style: 입출금내역 필터링 기능 마크업, css 적용

Commit:
feat: 월별/지출,수입별 리스트 필터 기능 추가

Commit:
chore: 서버 더미데이터 중 type 변경, id 타입 변경

Commit:
feat: store의 item 데이터를 서버의 데이터로 불러오는 최초 fetch 함수 추가

Commit:
feat: 삭제 기능 api 연결

Commit:
refactor: 이벤트 관련 함수 분리

Commit:
feat: 결제수단 api 연결

Commit:
chore: 오타 수정

Commit:
feat: 아이템 추가, 수정 api 연결

Commit:
refactor: 변수명 변경
income -> deposit, expense -> withdraw

Commit:
chore: date 관련 함수 폴더 위치 변경

Commit:
Merge branch 'feature/server-api' into dev

Commit:
chore: 필요없는 폴더 제거

Commit:
chore: 이벤트 등록 유틸 함수명 변경
기존에는 native API(addEventListener)와 이름이 같아 헷갈리는 문제가 발생하여 addEvent로 함수명을 변경했습니다.

Commit:
fix: 아이템 추가할 때 id의 type을 string으로 변경

Commit:
fix: 삭제버튼 핸들러 동작 이후 수정 핸들러도 동작하는 문제 해결

eunjin11 commented on src/components/transaction-form/TransactionFormView.js:
```diff
@@ -0,0 +1,198 @@
+  innerHTML += createDescription(description);
+  innerHTML += createMethodSelect({ method, methods });
+  innerHTML += createCategorySelect({ type, category });
+  innerHTML += createTransactionSubmitButton();
```
+=연산자를 이용해 문자열을 붙이면 문자열을 붙일 때 마다 재할당이 일어나니까
const innerHTML=createDateInput()+createAmount()~~ 같은 방식이나
const innerHTML=`${createDateInput()}${createAmount()}` 이런 방식이 더 낫지 않을까? 라는 생각이 들었습니당
코드가 좀 더 깔끔해보일 것 같기도 하구요!!

근데 템플릿 리터럴 방식을 사용하면 내부에서 파싱을 하는데 시간이 또 걸릴 거 같아서 성능 차이는 별로 없을 것 같기도 합니다.....

배열에 넣고 join으로 합치는 것도 괜찮을 것 같아요
배열을 join할 때는 전체 길이 미리 계산 → 버퍼 생성 → 한 번에 병합 과정을 거친다고 하는데 그럼 재할당을 덜 할 수 있지 않을까..? 하는 생각이..ㅎㅎ

yyeonzu replied:
헉 생각도 못해본 포인트네요... 내일 반영해보겠습니다!!

eunjin11 commented on src/store/FilterState.js:
```diff
@@ -0,0 +1,33 @@
+import { getCurrentYearMonth } from "../utils/date.js";
+import Observable from "./Observable.js";
+
+class FilterState extends Observable {
```
저도 filterState 옵저버 스토어로 저장해서 관리하고 싶었는데.. 너무 복잡해질 것 같아서 포기했어요 ㅜㅜ
filterStore 구현 멋있어요

yyeonzu replied:
월별로 데이터를 불러오는 파싱을 미루고 미루다.. 게으른 덕분에 필터를 옵저버로 관리하게 됐어요 ㅎㅎㅎ 은진님은 차트까지 다하셨으니,,, 낼 고고해보세요 😆

eunjin11 reviewed (commented):
store을 정말 잘쓰시는 것 같아요.. 멋있습니다

Commit:
feat: 라우팅 관련 옵저버 추가
navbarstate의 변경에 따라 Layout을 다시 그리는 옵저버 추가

Commit:
feat: 캘린더 기능 추가 (Observer, view, controller)

Commit:
styel: 캘린더 css 추가

Commit:
fix: 최초의 nav 상태가 url을 따르지 않고 언제나 메인으로 초기화되는 문제 해결

Commit:
refactor: 찾는 요소가 없을 경우 렌더링하지 않고  리턴하도록 가드 추가

Commit:
fix: 페이지를 불러올 때마다 옵저버가 중복 구독되는 문제 해결

기존에는 unsubscribe를 하지 않아 페이지를 create 할 때마다 옵저버가 중복 등록되었는데, 페이지 전환에 앞서 모든 옵저버를 해제하고 새로운 페이지를 불러올 수 있도록 코드를 수정했습니다.

removeObserver 함수를 구현하면서 addObserver 함수에서도 이미 옵저버가 있는지 확인하는 과정을 추가했습니다.

Commit:
chore: 디버깅 콘솔 제거

Commit:
refactor: 생성자의 date에 대해 오늘 날짜를 받지 않고, DateState를 받도록 변경

Commit:
refactor: 폼 제출 이후 날짜 초기화를 방금 제출한 날짜로 설정

Commit:
fix: 입력 폼 속 date input이 dateState가 아닌 formState를 구독하도록 변경

Commit:
refactor: DateInputObserver 생성자 코드 수정

Commit:
fix: DateState가 변경되면 FormState의 date가 변경될 수 있도록 데이터 흐름 수정
